README.txt
This extension provides an API that can be used in other extensions for parsing expressions
with a syntax very much like the getText functionality of TypoScript. Please refer to the
documentation for more details.

ChangeLog
2016-03-14 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Released to TER as version 2.0.0, resolves #75524

2016-03-14 Fabien Udriot  <fabien.udriot@ecodev.ch>

	* Added composer file, resolves #75053

2015-06-05 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Moved to namespaces, verified compatibility with TYPO3 CMS 7, resolves #67305
	* Moved documentation to reST, resolves #67314

2015-05-09 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Verified compatibility with TYPO3 CMS 6.2, resolves #58661
	* Released to TER as version 1.8.0, resolves #58662

2013-05-27 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Verified compatibility with TYPO3 6.1, resolves #48576
	* Released to TER as version 1.7.0

2013-03-26 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Corrected syntax error in function processing, resolves #46700
	* Verified compatibility with TYPO3 6.0, resolves #46698
	* Corrected erroneous examples in the manual, resolves #46701
	* Released to TER as version 1.6.0

2012-07-10 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Released to TER as version 1.5.0

2012-06-26 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Cleaned up some code, tested compatibility with TYPO3 4.7, raised requirements to TYPO3 4.5, resolves #38389

2012-01-11 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Small code cleanup

2011-04-04 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Added require statement to Fluid view helper, resolves #25740
	* Added support for inline notation, resolves #25742
	* Released to TER as version 1.4.0

2010-12-28 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Released to TER as version 1.3.0

2010-10-20 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Add new function "quoteStr" (mapped to t3lib_db::quoteStr())

2010-09-28 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Added proper handling of function processing when value is an array

2010-09-27 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Removed suggestion of extension "context"
	* Released to TER as version 1.2.3

2010-09-04 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Removed unnecessary usage of "context" extension hook

2010-08-25 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Corrected behavior of exceptions when using alternative expressions
	* Released to TER as version 1.2.2

2010-08-20 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Corrected abusive empty() check
	* Released to TER as version 1.2.1

2010-06-11 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Added new function key "env"
	* Corrected a bug with multiple colons in an expression
	* Released to TER as version 1.2.0

2010-06-02 Fabien Udriot  <fabien.udriot@ecodev.ch>

	* Added View Helper for Fluid
	* Added shortcut key "fe_user"

2010-05-05 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Corrected parser behavior in case of empty string values
	* Released to TER as version 1.1.2

2010-02-02 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Added new function "strftime"
	* Released to TER as version 1.1.1

2010-01-20 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Added function calls on expressions
	* Fixed broken support for subexpressions
	* Released to TER as version 1.1.0

2010-01-20 Fabien Udriot  <fabien.udriot@ecodev.ch>

	* Added shortcut key "plugin"

2009-10-04 Francois Suter (Cobweb)  <typo3@cobweb.ch>

	* Released to TER as version 1.0.0


Classes/ExpressionParser.php
<?php
namespace Cobweb\Expressions;

/*
 * This file is part of the TYPO3 CMS project.
 *
 * It is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License, either version 2
 * of the License, or any later version.
 *
 * For the full copyright and license information, please read the
 * LICENSE.txt file that was distributed with this source code.
 *
 * The TYPO3 project - inspiring people to share!
 */

use Cobweb\Expressions\Exception\Exception;
use Cobweb\Expressions\KeyProcessorInterface;
use Cobweb\Expressions\ValuePostProcessorInterface;
use TYPO3\CMS\Core\Utility\ArrayUtility;
use TYPO3\CMS\Core\Utility\GeneralUtility;

/**
 * Utility class to parse strings and evaluate them as expressions or for any subexpressions they may contain.
 *
 * @author Francois Suter (Cobweb) <typo3@cobweb.ch>
 * @package TYPO3
 * @subpackage tx_expressions
 */
class ExpressionParser {
	/**
	 * @var string Extension key
	 */
	public static $extKey = 'expressions';

	/**
	 * @var array Local variables stored in the parser
	 */
	public static $vars = array();

	/**
	 * @var array Additional values stored in the parser
	 */
	public static $extraData = array();

	/**
	 * Takes a string and checks for sub-strings inside curly braces {}.
	 *
	 * Each such substring will be evaluated and replaced.
	 * The resulting string is then evaluated or not, depending on the second argument.
	 *
	 * @param string $string The string to parse
	 * @param boolean $doEvaluation TRUE if string should be evaluated before being returned (default)
	 * @return string The parsed string
	 */
	public static function evaluateString($string, $doEvaluation = TRUE) {
		$matches = array();
		$numReplacements = preg_match_all('/(\{.*?\})/', $string, $matches, PREG_SET_ORDER);
		// If there was nothing to match or the matching failed, return the string as is
		$result = $string;
		if (!empty($numReplacements)) {
			$searches = array();
			$replacements = array();
			foreach ($matches as $aMatch) {
				$searches[] = $aMatch[1];
				$expression = substr($aMatch[1], 1, strlen($aMatch[1]) - 2);
				$evaluatedExpression = $expression;
				try {
					$evaluatedExpression = self::evaluateExpression($expression);
				}
				catch (Exception $e) {
					if (TYPO3_DLOG) {
						GeneralUtility::devLog('Bad subexpression: ' . $expression . ' (' . $e->getMessage() . ')', 'expressions', 2);
					}
				}
				$replacements[] = $evaluatedExpression;
			}
			$result = str_replace($searches, $replacements, $string);
		}
		// Evaluate the string, if necessary
		$finalString = $result;
		if ($doEvaluation) {
			try {
				$finalString = self::evaluateExpression($result);
			}
			// If the evaluation fails, return the original string
			catch (Exception $e) {
				if (TYPO3_DLOG) {
					GeneralUtility::devLog('Could not evaluate string: ' . $result . ' (' . $e->getMessage() . ')', 'expressions', 2);
				}
			}
		}
		return $finalString;
	}

	/**
	 * Evaluates the value of a given expression.
	 *
	 * The expected syntax of a filter value is key:index1|index2|...
	 * Simple values will be used as is.
	 *
	 * @param string $expression The expression to evaluate
	 * @throws Exception
	 * @return string The value for the filter
	 */
	public static function evaluateExpression($expression) {
		$returnValue = '';
		$hasValue = FALSE;
		if (!isset($expression) || $expression === '') {
			throw new Exception('Empty expression received');
		} else {
			// First of all, evaluate any subexpressions that may be contained in the expression
			$parsedExpression = self::evaluateString($expression, FALSE);
			// An expression may contain several expressions as alternativtye values, separated by a double slash (//)
			$allExpressions = GeneralUtility::trimExplode('//', $parsedExpression, TRUE);
			$numberOfAlternatives = count($allExpressions);
			foreach ($allExpressions as $anExpression) {
				// Decrease the number of remaining alternatives
				$numberOfAlternatives--;
				// Check if there's a function call
				$functions = array();
				if (strpos($anExpression, '->') !== FALSE) {
					// Split on the function call marker (->)
					$expressionParts = GeneralUtility::trimExplode('->', $anExpression, TRUE);
					// The first part is the expression itself
					$anExpression = array_shift($expressionParts);
					// All other parts are function definitions
					$functions = $expressionParts;
				}
				// If there's no colon (:) in the expression, take it to be a literal value and return it as is
				if (strpos($anExpression, ':') === FALSE) {
					$returnValue = $anExpression;
					$hasValue = TRUE;
				} else {
					$subParts = GeneralUtility::trimExplode(':', $anExpression, TRUE);
					$key = array_shift($subParts);
					$indices = implode(':', $subParts);
					if (empty($indices)) {
						throw new Exception('No indices in expression: ' . $expression);
					}
					$key = strtolower($key);
					switch ($key) {
						// Search for a value in the TSFE
						case 'tsfe':
							if (TYPO3_MODE == 'FE') {
								try {
									$returnValue = self::getValue($GLOBALS['TSFE'], $indices);
									$hasValue = TRUE;
								}
								catch (Exception $e) {
									continue;
								}
							} else {
								// Throw exception, but only if we have run out of alternatives
								if ($numberOfAlternatives == 0) {
									throw new Exception('TSFE not available in this mode (' . TYPO3_MODE . ')');
								}
							}
							break;
						// Search for a value in the page record
						// This is a convenience shortcut, as the page record is in the TSFE
						case 'page':
							if (TYPO3_MODE == 'FE') {
								try {
									$returnValue = self::getValue($GLOBALS['TSFE']->page, $indices);
									$hasValue = TRUE;
								}
								catch (Exception $e) {
									continue;
								}
							} else {
								// Throw exception, but only if we have run out of alternatives
								if ($numberOfAlternatives == 0) {
									throw new Exception('TSFE->page not available in this mode (' . TYPO3_MODE . ')');
								}
							}
							break;
						// Search for a value in the template configuration
						// This is a convenience shortcut, as the template configuration is in the TSFE
						case 'config':
							if (TYPO3_MODE == 'FE') {
								try {
									$returnValue = self::getValue($GLOBALS['TSFE']->config['config'], $indices);
									$hasValue = TRUE;
								}
								catch (Exception $e) {
									continue;
								}
							} else {
								// Throw exception, but only if we have run out of alternatives
								if ($numberOfAlternatives == 0) {
									throw new Exception('TSFE->config not available in this mode (' . TYPO3_MODE . ')');
								}
							}
							break;
						// Search for a value in the merged GET and POST arrays
						case 'plugin':
							if (TYPO3_MODE == 'FE') {
								try {
									$returnValue = self::getValue($GLOBALS['TSFE']->tmpl->setup['plugin.'], $indices);
									$hasValue = TRUE;
								}
								catch (Exception $e) {
									continue;
								}
							} else {
								// Throw exception, but only if we have run out of alternatives
								if ($numberOfAlternatives == 0) {
									throw new Exception('Plugin setup not available in this mode (' . TYPO3_MODE . ')');
								}
							}
							break;
						// Search for a value in the merged GET and POST arrays
						case 'gp':
							try {
								$returnValue = self::getValue(
									array_merge(
										GeneralUtility::_GET(),
										GeneralUtility::_POST()
									),
									$indices
								);
								$hasValue = TRUE;
							}
							catch (Exception $e) {
								continue;
							}
							break;
						// Search for a value in the local variables
						case 'vars':
							try {
								$returnValue = self::getValue(self::$vars, $indices);
								$hasValue = TRUE;
							}
							catch (Exception $e) {
								continue;
							}
							break;
						// Search for a value in the extra data
						case 'extra':
							try {
								$returnValue = self::getValue(self::$extraData, $indices);
								$hasValue = TRUE;
							}
							catch (Exception $e) {
								continue;
							}
							break;
						// Calculate a value using the PHP date() function
						case 'date':
							$returnValue = date($indices);
							$hasValue = TRUE;
							break;
						case 'strtotime':
							$value = strtotime($indices);
							if ($value === FALSE || $value === -1) {
								throw new Exception('Date string could not be parsed: ' . $indices);
							} else {
								$returnValue = $value;
								$hasValue = TRUE;
							}
							break;
						// Get data from the session
						// The session key is the first segment after the "session" keyword
						case 'session':
							$segments = GeneralUtility::trimExplode('|', $indices, TRUE);
							$cacheKey = array_shift($segments);
							$indices = implode('|', $segments);
							$cache = $GLOBALS['TSFE']->fe_user->getKey('ses', $cacheKey);
							if (empty($cache)) {
								// Throw exception, but only if we have run out of alternatives
								if ($numberOfAlternatives == 0) {
									throw new Exception('No session data found for expression: ' . $expression);
								} else {
									continue;
								}
							}
							try {
								$returnValue = self::getValue($cache, $indices);
								$hasValue = TRUE;
							}
							catch (Exception $e) {
								continue;
							}
							break;
						// Search for a value in FE User
						case 'fe_user':
							if (TYPO3_MODE == 'FE') {
								try {
									$returnValue = self::getValue($GLOBALS['TSFE']->fe_user->user, $indices);
									$hasValue = TRUE;
								}
								catch (Exception $e) {
									continue;
								}
							} else {
								// Throw exception, but only if we have run out of alternatives
								if ($numberOfAlternatives == 0) {
									throw new Exception('TSFE->fe_user not available in this mode (' . TYPO3_MODE . ')');
								}
							}
							break;
						// Search for a value in the environment variables as returned by GeneralUtility::getIndpEnv()
						case 'env':
							$returnValue = GeneralUtility::getIndpEnv(strtoupper($indices));
							$hasValue = TRUE;
							break;
							// If none of the standard keys matched, try looking for a hook for that given key
						default:
							if (isset($GLOBALS['TYPO3_CONF_VARS']['EXTCONF'][self::$extKey]['keyProcessor'][$key])) {
								/** @var $keyProcessor KeyProcessorInterface */
								$keyProcessor = GeneralUtility::getUserObj($GLOBALS['TYPO3_CONF_VARS']['EXTCONF'][self::$extKey]['keyProcessor'][$key]);
								if ($keyProcessor instanceof KeyProcessorInterface) {
									$returnValue = $keyProcessor->getValue($indices);
									$hasValue = TRUE;
								}
							}
							break;
					}
				}
				// If a value was found, process it and exit the loop
				if ($hasValue) {
					// Call functions, if any
					if (count($functions) > 0) {
						foreach ($functions as $functionDefinition) {
							try {
								$returnValue = self::processFunctionCall($returnValue, $functionDefinition);
							}
							// Do nothing on exceptions, value is unchanged
							catch (Exception $e) {
									continue;
							}
						}
					}
					break;
				}
			}
		}
		// If a value was found, call a post-processing hook and return the value
		if ($hasValue) {
			if (is_array($GLOBALS['TYPO3_CONF_VARS']['EXTCONF'][self::$extKey]['postprocessReturnValue'])) {
				foreach($GLOBALS['TYPO3_CONF_VARS']['EXTCONF'][self::$extKey]['postprocessReturnValue'] as $className) {
						/** @var $postProcessor ValuePostProcessorInterface */
					$postProcessor = GeneralUtility::getUserObj($className);
					if ($postProcessor instanceof ValuePostProcessorInterface) {
						$returnValue = $postProcessor->postprocessReturnValue($returnValue);
					}
				}
			}
			return $returnValue;

		// No value was found, throw an exception
		} else {
			throw new Exception('No value found for expression: ' . $expression);
		}
	}

	/**
	 * Gets a value from inside a multi-dimensional array or object.
	 *
	 * NOTE: this code is largely inspired by \TYPO3\CMS\Frontend\ContentObject\ContentObjectRenderer::getGlobal().
	 *
	 * @param mixed $source Array or object to look into
	 * @param string $indices "Path" of indices inside the multi-dimensional array, of the form index1|index2|...
	 * @throws Exception
	 * @return mixed Whatever value was found in the array, but it should be a simple type
	 */
	protected static function getValue($source, $indices) {
		$value = $source;
		if (empty($indices)) {
			throw new Exception('No key given for source');
		} else {
			$indexList = GeneralUtility::trimExplode('|', $indices, TRUE);
			foreach ($indexList as $key) {
				if (is_object($value) && isset($value->$key) && $value->$key !== '') {
					$value = $value->$key;
				} elseif (is_array($value) && isset($value[$key]) && $value[$key] !== '') {
					$value = $value[$key];
				} else {
					throw new Exception('Key ' . $indices . ' not found in source');
				}
			}
		}
		return $value;
	}

	/**
	 * Handles the function calls that can be made inside expressions.
	 *
	 * NOTE: if the function cannot be called, the original value is returned untouched.
	 *
	 * @param mixed $value The value to pass to the function (the result of an expression)
	 * @param string $functionDefinition The definition of the function to call in the appropriate syntax
	 * @return mixed The value, as processed by the function
	 */
	protected function processFunctionCall($value, $functionDefinition) {
		// Initializations
		$arguments = array();
		// Separate function key and list of arguments
		list($function, $argumentsString) = GeneralUtility::trimExplode(':', $functionDefinition, TRUE);
		// Get arguments as array
		if (isset($argumentsString) && $argumentsString != '') {
			$arguments = GeneralUtility::trimExplode(',', $argumentsString, TRUE);
		}
		// Execute the function, recursively if value is an array
		if (is_array($value)) {
			$processedValue = self::executeFunctionOnArray($value, $function, $arguments);
		} else {
			$processedValue = self::executeFunctionOnItem($value, $function, $arguments);
		}
		return $processedValue;
	}

	/**
	 * Loops on all items in an array and executes the function on it
	 * (or calls itself recursively if item is itself an array).
	 *
	 * @param array $value List of values to call the function on
	 * @param string $function The key of the function to call
	 * @param array $arguments The list of arguments to pass in the function call
	 * @return array The processed array of values
	 */
	protected function executeFunctionOnArray(array $value, $function, $arguments) {
		foreach ($value as $key => $item) {
			if (is_array($item)) {
				$value[$key] = self::executeFunctionOnArray($item, $function, $arguments);
			} else {
				$value[$key] = self::executeFunctionOnItem($item, $function, $arguments);
			}
		}
		return $value;
	}

	/**
	 * Calls the processing function on a given value.
	 *
	 * @param mixed $value Value to call the function on
	 * @param string $function The key of the function to call
	 * @param array $arguments The list of arguments to pass in the function call
	 * @throws Exception
	 * @return array The processed values
	 */
	protected function executeFunctionOnItem($value, $function, $arguments) {
		$processedValue = $value;
			// Execute function
		switch ($function) {
			case 'fullQuoteStr':
				if (count($arguments) < 1) {
					throw new Exception('fullQuoteStr() requires 1 argument (table name)');
				} elseif (!isset($GLOBALS['TYPO3_DB'])) {
					throw new Exception('TYPO3_DB object not available');
				} else {
					$processedValue = self::getDatabaseConnection()->fullQuoteStr($value, $arguments[0]);
				}
				break;
			case 'quoteStr':
				if (count($arguments) < 1) {
					throw new Exception('quoteStr() requires 1 argument (table name)');
				} elseif (!isset($GLOBALS['TYPO3_DB'])) {
					throw new Exception('TYPO3_DB object not available');
				} else {
					$processedValue = self::getDatabaseConnection()->quoteStr($value, $arguments[0]);
				}
				break;
			case 'strip_tags':
				$functionArguments = array($value);
				if (isset($arguments[0])) {
					$functionArguments[] = $arguments[0];
				}
				$processedValue = call_user_func_array('strip_tags', $functionArguments);
				break;
			case 'removeXSS':
				$processedValue = GeneralUtility::removeXSS($value);
				break;
			case 'intval':
				$functionArguments = array($value);
				if (isset($arguments[0])) {
					$functionArguments[] = $arguments[0];
				}
				$processedValue = call_user_func_array('intval', $functionArguments);
				break;
			case 'floatval':
				$processedValue = floatval($value);
				break;
			case 'boolean':
				$processedValue = !empty($value);
				break;
			case 'hsc':
				$functionArguments = array($value);
				if (count($arguments) > 0) {
					foreach ($arguments as $item) {
						$functionArguments[] = $item;
					}
				}
				$processedValue = call_user_func_array('htmlspecialchars', $functionArguments);
				break;
			case 'strftime':
				if (count($arguments) < 1) {
					throw new Exception('strftime() requires 1 argument (date format)');
				} else {
					$processedValue = strftime($arguments[0], $value);
				}
				break;

			// If no standard key matches, try to call user-defined function
			default:
				if (isset($GLOBALS['TYPO3_CONF_VARS']['EXTCONF'][self::$extKey]['customFunction'][$function])) {
					$functionArguments = array($value);
					if (count($arguments) > 0) {
						foreach ($arguments as $item) {
							$functionArguments[] = $item;
						}
					}
					// Call user function
					// Note the last parameter: since this class is purely static,
					// a reference to it cannot be passed to the user function,
					// so we instantiate a dummy object instead
					$processedValue = GeneralUtility::callUserFunction(
						$GLOBALS['TYPO3_CONF_VARS']['EXTCONF'][self::$extKey]['customFunction'][$function],
						$functionArguments,
						new \stdClass
					);
				}
				break;
		}
		return $processedValue;
	}

	/**
	 * This method can be used to store whatever local variables make sense into the parser, for later retrieval
	 * In the case of a FE plugin, it would be the piVars
	 *
	 * @param array $vars Array of values
	 * @param boolean $reset TRUE if self::$vars must be reset
	 * @return void
	 */
	public static function setVars($vars, $reset = FALSE) {
		if (is_array($vars)) {
			if ($reset) {
				self::$vars = $vars;
			} else {
				ArrayUtility::mergeRecursiveWithOverrule(self::$vars, $vars);
			}
		}
	}

	/**
	 * Stores additional variables into the parser, that should not be mixed up
	 * with the local variables stored in self::$vars.
	 *
	 * @param array	$data Array of values
	 * @param boolean $reset TRUE if self::$extraData must be reset
	 * @return void
	 * @see \Cobweb\Expressions\Parser::setVars()
	 */
	public static function setExtraData($data, $reset = FALSE) {
		if (is_array($data)) {
			if ($reset) {
				self::$extraData = $data;
			} else {
				ArrayUtility::mergeRecursiveWithOverrule(self::$extraData, $data);
			}
		}
	}

	/**
	 * Returns the global database object.
	 *
	 * @return \TYPO3\CMS\Core\Database\DatabaseConnection
	 */
	protected static function getDatabaseConnection() {
		return $GLOBALS['TYPO3_DB'];
	}
}


Classes/KeyProcessorInterface.php
<?php
namespace Cobweb\Expressions;

/*
 * This file is part of the TYPO3 CMS project.
 *
 * It is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License, either version 2
 * of the License, or any later version.
 *
 * For the full copyright and license information, please read the
 * LICENSE.txt file that was distributed with this source code.
 *
 * The TYPO3 project - inspiring people to share!
 */

/**
 * Interface which defines the method to implement when creating a special key processor for expressions.
 *
 * @author Francois Suter (Cobweb) <typo3@cobweb.ch>
 * @package TYPO3
 * @subpackage tx_expressions
 */
interface KeyProcessorInterface {
	/**
	 * This method must be implemented to handle the "indices" part it receives from the parser.
	 *
	 * Assuming an expression such as:
	 *
	 * mykey:foo|bar
	 *
	 * the method will receive the string foo|bar.
	 * It is expected to return a value that makes sense or else throw an exception.
	 *
	 * @param string $indices The string to be interpreted
	 * @return mixed The resulting value
	 */
	public function getValue($indices);
}


Classes/Sample/FunctionProcessor.php
<?php
namespace Cobweb\Expressions\Sample;

/*
 * This file is part of the TYPO3 CMS project.
 *
 * It is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License, either version 2
 * of the License, or any later version.
 *
 * For the full copyright and license information, please read the
 * LICENSE.txt file that was distributed with this source code.
 *
 * The TYPO3 project - inspiring people to share!
 */

/**
 * Sample class providing an example user function for post-processing expressions.
 *
 * To activate that function, enter the following is some localconf file:
 *
 * $GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['expressions']['customFunction']['offset'] = 'Cobweb\\Expressions\\Sample\\FunctionProcessor->offset';
 *
 * To use it in an expression, enter something like:
 *
 * gp:foo|bar->offset:10
 *
 * This will take the foo[bar] GET/POST variable and offset it by 10
 *
 * @author Francois Suter (Cobweb) <typo3@cobweb.ch>
 * @package TYPO3
 * @subpackage tx_expressions
 */
class FunctionProcessor {
	/**
	 * Sample method to process expressions using the function-calling syntax
	 * i.e. this method might be called using:
	 *
	 * gp:foo|bar->offset:10
	 *
	 * The method receives an array of parameters and a reference to a dummy object
	 * (there's no real reference to a calling object, because the calling object is static).
	 * The array of parameters always contains the value from the expression first
	 * (i.e. in $parameters[0]).
	 * Any additional parameters required by the function come next. In this sample's
	 * case, the second parameter ($parameters[1]) is a number by which the value
	 * will be offset.
	 *
	 * The method *must* return the expression's value, even if unchanged.
	 *
	 * @param array	$parameters Indexed list of parameters. The first one is the expression's result.
	 * @param object $parentObject Dummy object. Not usable.
	 * @return mixed The processed value
	 */
	public function offset($parameters, $parentObject) {
		$value = intval($parameters[0]);
		$offset = 0;
		if (isset($parameters[1])) {
			$offset = intval($parameters[1]);
		}
		return $value + $offset;
	}
}


Classes/ValuePostProcessorInterface.php
<?php
namespace Cobweb\Expressions;

/*
 * This file is part of the TYPO3 CMS project.
 *
 * It is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License, either version 2
 * of the License, or any later version.
 *
 * For the full copyright and license information, please read the
 * LICENSE.txt file that was distributed with this source code.
 *
 * The TYPO3 project - inspiring people to share!
 */

/**
 * Interface which defines the method to implement when creating a hook to post-process parsed expressions.
 *
 * @author Francois Suter (Cobweb) <typo3@cobweb.ch>
 * @package TYPO3
 * @subpackage tx_expressions
 */
interface ValuePostProcessorInterface {
	/**
	 * This method must be implemented for post-processing a parsed value.
	 *
	 * It must return a value too, even if that is the unchanged input value.
	 *
	 * @param mixed $value The value that resulted from parsing an expression
	 * @return mixed The resulting value
	 */
	public function postprocessReturnValue($value);
}


Classes/ViewHelpers/EvaluateViewHelper.php
<?php
namespace Cobweb\Expressions\ViewHelpers;

/*
 * This file is part of the TYPO3 CMS project.
 *
 * It is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License, either version 2
 * of the License, or any later version.
 *
 * For the full copyright and license information, please read the
 * LICENSE.txt file that was distributed with this source code.
 *
 * The TYPO3 project - inspiring people to share!
 */

use Cobweb\Expressions\ExpressionParser;
use TYPO3\CMS\Fluid\Core\ViewHelper\AbstractViewHelper;

/**
 * This class is a Expressions view helper for the Fluid templating engine.
 *
 * In TypoScript templates there exists a function called “getText”, which makes it possible
 * to retrieve values from a great number of sources (GET/POST, TSFE, rootline, etc.).
 * Extension "expressions" provides a library which aims to reproduce this capability,
 * so that it can be used by developers where Typoscript is not used for whatever reasons.
 *
 * Like for the TypoScript function, an expression – at a minimum – is comprised of a key, a colon (:) and a value.
 * Example: tsfe:id
 *
 * Expression keys reference:
 *
 * tsfe:      get a value from the $TSFE global variable
 * page:      get a value related to the current page, as stored in $TSFE->page
 * config:    get a value from the “config” object, as stored in $TSFE->config['config']
 * plugin:    get a TypoScript property for a plugin as stored in $GLOBALS['TSFE']->tmpl->setup['plugin.'].
 * gp:        get a value from either the $_GET or $_POST superglobal arrays
 * vars:      get a value from "internal" variables.
 * extra:     see documentation
 * date:      get values related to the current time, using formats from the PHP date() function.
 * strtotime: get a timestamp by interpreting a human-readable date.
 * session:   get values from some structure stored in the temporary session
 * fe_user:   get values from FE User
 *
 * More documentation can be found at
 * http://typo3.org/extensions/repository/view/expressions/current/
 *
 * = Examples =
 *
 * Assuming that namespace has been declared as follows:
 * {namespace expression = Cobweb\Expressions\ViewHelpers}
 *
 * <code title="Simple string with expression to evaluate">
 * <expression:evaluate>Current page id is \{tsfe:id\}</expression:evaluate>
 * </code>
 * <output>
 * Current page id is 1
 * (assuming the current page has uid = 1, of course)
 * </output>
 * Note the escaped curly braces, so that Fluid is not confused
 *
 * <code title="Inline notation">
 * {expression:evaluate(expression:'Current user is \{fe_user:username\}')}
 * </code>
 * <output>
 * Current user is zaphod
 * (assuming the current FE user's username is "zaphod")
 * </output>
 *
 * @package TYPO3
 * @subpackage tx_expressions
 * @author Fabien Udriot <fabien.udriot@ecodev.ch>
 * @author Francois Suter <typo3@cobweb.ch>
 */
class EvaluateViewHelper extends AbstractViewHelper {

	/**
	 * Evaluates expression throughout the Expression Parser.
	 *
	 * @param string $expression Expression to be evaluated
	 * @return string the evaluated string.
	 */
	public function render($expression = NULL) {
		if ($expression === NULL) {
			$expression = $this->renderChildren();
		};
		// Replace escaped curly braces
		$searches[] = '\{';
		$replaces[] = '{';
		$searches[] = '\}';
		$replaces[] = '}';
		$expression = str_replace($searches, $replaces, $expression);
		// Evaluate and return
		return ExpressionParser::evaluateString($expression);
	}

}


ext_emconf.php
<?php

/*********************************************************************
* Extension configuration file for ext "expressions".
*
* Generated by ext 11-04-2016 16:08 UTC
*
* https://github.com/t3elmar/Ext
*********************************************************************/

$EM_CONF[$_EXTKEY] = array (
  'title' => 'Generic expression parser',
  'description' => 'A library for bringing into extensions something like the TypoScript getText function.',
  'category' => 'misc',
  'author' => 'Francois Suter (Cobweb)',
  'author_email' => 'typo3@cobweb.ch',
  'state' => 'stable',
  'uploadfolder' => 0,
  'createDirs' => '',
  'clearCacheOnLoad' => 0,
  'author_company' => '',
  'version' => '2.0.0',
  'constraints' => 
  array (
    'depends' => 
    array (
      'typo3' => '7.0.0-7.99.99',
    ),
    'conflicts' => 
    array (
    ),
    'suggests' => 
    array (
    ),
  ),
  '_md5_values_when_last_written' => 'a:29:{s:9:"ChangeLog";s:4:"10bf";s:10:"README.txt";s:4:"9be0";s:13:"composer.json";s:4:"4213";s:12:"ext_icon.png";s:4:"cdfa";s:28:"Classes/ExpressionParser.php";s:4:"b901";s:33:"Classes/KeyProcessorInterface.php";s:4:"7970";s:39:"Classes/ValuePostProcessorInterface.php";s:4:"c049";s:31:"Classes/Exception/Exception.php";s:4:"9d9b";s:36:"Classes/Sample/FunctionProcessor.php";s:4:"8401";s:42:"Classes/ViewHelpers/EvaluateViewHelper.php";s:4:"be8b";s:26:"Documentation/Includes.txt";s:4:"c83c";s:23:"Documentation/Index.rst";s:4:"dbf1";s:26:"Documentation/Settings.yml";s:4:"4bf7";s:33:"Documentation/Developer/Index.rst";s:4:"dbdc";s:37:"Documentation/Developer/Api/Index.rst";s:4:"2b57";s:44:"Documentation/Developer/Exceptions/Index.rst";s:4:"c6b4";s:39:"Documentation/Developer/Hooks/Index.rst";s:4:"61f7";s:46:"Documentation/Developer/NewFunctions/Index.rst";s:4:"4929";s:41:"Documentation/Developer/NewKeys/Index.rst";s:4:"7f4e";s:42:"Documentation/Developer/Security/Index.rst";s:4:"f277";s:43:"Documentation/Developer/Variables/Index.rst";s:4:"30f5";s:44:"Documentation/Developer/ViewHelper/Index.rst";s:4:"5965";s:36:"Documentation/Installation/Index.rst";s:4:"dd3c";s:36:"Documentation/Introduction/Index.rst";s:4:"67bc";s:37:"Documentation/KnownProblems/Index.rst";s:4:"056c";s:28:"Documentation/User/Index.rst";s:4:"73fc";s:43:"Documentation/User/ExpressionKeys/Index.rst";s:4:"74eb";s:45:"Documentation/User/ExpressionSyntax/Index.rst";s:4:"4ff9";s:41:"Documentation/User/FunctionKeys/Index.rst";s:4:"15d8";}',
  'comment' => 'Updated for TYPO3 CMS 7 LTS, moved to namespaces',
  'user' => 'francois',
);

?>